transformer_engine/jax/mlp.py

```diff
@@ -633,36 +341,62 @@ def _layernorm_gelu_fp8_mlp_bwd_rule(
 
     gemm1_x_idx, gemm1_kernel_idx, gemm1_grad_idx = FP8Helper.get_fp8_meta_indices(0)
 
-    dgelu_amax = amax[gemm1_grad_idx, 0:1]
-    dgelu_scale = scale[gemm1_grad_idx]
-    dgelu_scale_inv = scale_inv[gemm1_grad_idx]
-
-    casted_dgelu, casted_dgelu_t, dbias_1, updated_dgelu_amax = dgelu_dbias_cast_transpose(
-        dgrad_2,
-        dot_1_output,
-        dgelu_amax,
-        dgelu_scale,
-        dgelu_scale_inv,
-        bwd_dtype,
-        static_axis_boundary=-1,
-        transpose_axis_boundary=-1)
-
-    dbias_1 = jnp.reshape(dbias_1, bias_1_shape)
+    dactivation_lu_amax = amax[gemm1_grad_idx, 0:1]
+    dactivation_lu_scale = scale[gemm1_grad_idx]
+    dactivation_lu_scale_inv = scale_inv[gemm1_grad_idx]
+
+    dactivation_lu_cast_transpose = activation_lu_fp8_dict[activation_type]["bwd"]
+    if not is_gated and use_bias:
+        casted_dactivation_lu, casted_dactivation_lu_t, dbias_1, updated_dactivation_lu_amax = \
+        dactivation_lu_cast_transpose(
+            dgrad_2,
+            dot_1_output,
+            dactivation_lu_amax,
+            dactivation_lu_scale,
+            dactivation_lu_scale_inv,
+            bwd_dtype,
+            static_axis_boundary=-1,
+            transpose_axis_boundary=-1)
+
+        dbias_1 = jnp.reshape(dbias_1, bias_1_shape)
+    elif is_gated and not use_bias:
+        casted_dactivation_lu, casted_dactivation_lu_t, updated_dactivation_lu_amax = \
+        dactivation_lu_cast_transpose(
+            dgrad_2,
+            dot_1_output,
+            dactivation_lu_amax,
+            dactivation_lu_scale,
+            dactivation_lu_scale_inv,
+            bwd_dtype,
+            static_axis_boundary=-1)
+    else:
+        raise NotImplementedError
```

Comment:
@zlsh80826 @mingxu1067 @ptrendx Is there a reason why we can't support fused biases here via `dactivation() -> cast_transpose_dbias()`?

If we ultimately don't add partially fused biases into the underlying custom call, then we should probably have a useful error message here stating that unfused biases are supported for all activations via `te.flax.LayerNormMLP`.

Reply:
@mingxu1067 This is my ignorance with XLA's internal mechanisms but I'm surprised that it's more performant to do `LayerNormDenseGeneral w/ biases-> Activation -> DenseGeneral w/biases` and let JAX autodiff handle the backward pass than to implement a custom BWD with `dactivation() -> cast_transpose_dbias()`. Does XLA construct a better fused kernel in for this particular autodiff'd BWD than what we can do ourselves?

In any case, it should be pretty easy to template the `dactivation_dbias_cast_transpose()` kernel in the next PR to address this without running into any potential issues with separate `dactivation() -> cast_transpose_dbias()` calls.

Reply:
> Does XLA construct a better fused kernel in for this particular autodiff'd BWD than what we can do ourselves?
As I know, XLA does not do well on `cast_transpose` related fusion. IIRC, the perf is about 50% slower than TE's custom calls